src/app/shared/asyncapi.service.ts

```diff
@@ -82,43 +82,47 @@ export class AsyncApiService {
         };
     }
 
-    private mapServers(servers: { [key: string]: Server }): Map<string, Server> {
+    private mapServers(servers: ServerAsyncApi["servers"]): Map<string, Server> {
         const s = new Map<string, Server>();
         Object.entries(servers).forEach(([k, v]) => s.set(k, v));
         return s;
     }
 
-    private mapChannels(channels: {
-        [key: string]: {
-            description?: string;
-            subscribe?: {
-                message: Message;
-                bindings?: any;
-            };
-            publish?: {
-                message: Message | { oneOf: Message[] };
-                bindings?: any;
-            };
-        }
-    }): Channel[] {
+    private mapChannels(channels: ServerAsyncApi["channels"]): Channel[] {
         const s = new Array<Channel>();
         Object.entries(channels).forEach(([k, v]) => {
-            let operation = v.publish ? v.publish : v.subscribe;
-            let isSubscribe = !!v.subscribe;
+            const subscriberChannels = this.mapChannel(k, v.description, v.subscribe, " consumer")
+            subscriberChannels.forEach(channel => s.push(channel))
 
-            let messages: Message[] = 'oneOf' in operation.message ? operation.message.oneOf : [operation.message];
-            messages.forEach(message => s.push({
-                name: k,
-                description: v.description,
-                operation: this.mapOperation(isSubscribe, message, operation.bindings)
-            }))
+            const publisherChannels = this.mapChannel(k, v.description, v.publish, " producer")
+            publisherChannels.forEach(channel => s.push(channel))
         });
         return s;
     }
 
-    private mapOperation(isSubscribe: boolean, message: Message, bindings?: any): Operation {
+    private mapChannel(
+        topicName: string,
+        description: ServerAsyncApi["channels"][""]["description"],
+        operation: ServerAsyncApi["channels"][""]["subscribe"] | ServerAsyncApi["channels"][""]["publish"],
```

Comment:
Ok I'm sure this works but could you please explain what this means? I'm not familiar with this syntax

Reply:
This is the same syntax as above.

Similar as if you access nested properties on a variable, here the nested type is accessed.
Since there is so much nesting, this looks a bit clunky. An alternative would to split each nested type up into its own interface, like AsyncServerApiChannelItem, ...

If you use an IDE, you should be able to click and jump to the definition by hovering.

Reply:
If you are talking about the operation parameter, that uses the union operator.
So the variable is either of type a or b. Typescript makes sure that only properties can be accessed that are existing on both